### ISSUE TITLE
util::psbt::map::Input can refer to 2 outputs

### ISSUE THREAD

**Issue by hashmap** [rust-bitcoin/rust-bitcoin#1366]
I guess I understand why but it's a part of public API and looks fragile, it should be no possibility to construct such invalid input

https://github.com/rust-bitcoin/rust-bitcoin/blob/d8e94cf1817699cc1de0cd692fed7f09067df29a/bitcoin/src/util/psbt/map/input.rs#L75-L84

It could be smth like 
```rust
pub(in crate::util::psbt) non_witness_utxo: Option<Transaction>
```
plus methods to expose those fields in a controlled way, however I see that we rely on `..Default::default()` in examples 

https://github.com/rust-bitcoin/rust-bitcoin/blob/master/bitcoin/examples/taproot-psbt.rs#L443-L456 

which would not work with private fields. I'd still prefer to sacrifice convenience of defaults.

**Comment by tcharding** [comment]
Thanks for raising the issue.

The PSBT structs are coded up against the spec in [BIP-174](https://en.bitcoin.it/wiki/BIP_0174). AFAIU a reader of that bip would expect to find these two fields public.

**Comment by sanket1729** [comment]
Indeed, those parts of public API are fragile. One possible way to address that might be to create. Ignoring the names, this might look like.

```
pub spent_utxo: Option<SpentUtxo>,

...
pub enum SpentUtxo {
	WitnessUtxoOnly(TxOut),
	NonWitnessUtxoOnly(Transaction),
	Both(Transaction), // This contains the information about the UTXO
}
```
This should enough information to correctly roundtrip psbt from/to hex/base64. (I think, without looking at the code) this is something we explicitly check in rust-miniscript so I think it makes sense to enforce this psbt API. This won't play well without psbt macros, but that is an implementation detail.

The plus side here is that we also maintain the ergonomic `..Deafult::default` way of doing things without which specifying all psbt fields is not fun. 

That being said, this is awkward for the same reason that @tcharding mentioned. It would be strange to have `witness_utxo()` and `non_witness_utxo()` APIs, while everything else is a public field. It breaks the most intuitive to program these things. The counterpoint here would if downstream implementations have to do this awkward check, we should reflect it in our API. 

All psbt applications use these fields, so it would break every single piece of code and require manual fixing. Not really sure how to evaluate this, would like to see what others think of this breaking change. 
For now, I am leaning toward fixing it with the above approach. But given the complexity of our existing macros dealing with all fields independently, this simple task might be hard to do. I guess my stance is, we should fix it, but its output might not be worth the effort for this.

**Comment by Kixunil** [comment]
There is actually more things that can be inconsistent. You may have an incorrect transaction there, you may have different number of `Input` and `TxIn`, same with `TxOut`, I bet we could find other things...

I dealt with this kind of problem in the past and came to conclusion that one should have some low-level structures intended for parsing and high-level ones for validation or checked construction. Such design works really well in my experience.

The `Psbt` we provide is such low-level structure but it's not documented as such. I think we should document it and maybe try to think of a safer higher-level interface. But note that it will be **very** difficult because of many invariants and many fields (including custom) in PSBT. I suspect the best thing is for each application to define their on.

**Comment by sanket1729** [comment]
> I think we should document it and maybe try to think of a safer higher-level interface.

Agreed on this. Some of these things become easier with psbtv2, but there could be some other invariants there.

**Comment by hashmap** [comment]
Enum way was my first idea but it felt a bit alien in this low-level structure which follows the BIP as @tcharding mentioned. Plus existing macros expect a type, not an enum variant.
I agree with @Kixunil I think the main issue is that a low-level abstraction is directly exposed in API.

**Comment by Upasana33** [comment]
Hey, I am interested in working on this issue

**Comment by Kixunil** [comment]
@Upasana33 I assume you mean adding documentation, go ahead!

**Comment by Upasana33** [comment]
Thank you so much for assigning me @Kixunil. I will get started right away, and keep you posted.
